rstar/src/algorithm/removal.rs

```diff
@@ -18,43 +18,70 @@ where
     Params: RTreeParams,
     R: SelectionFunction<T>,
 {
-    let mut result = None;
+    remove_recursive::<_, Params, _>(node, removal_function, false).pop()
+}
+
+pub fn remove_all<T, Params, R>(node: &mut ParentNode<T>, removal_function: &R) -> Vec<T>
+where
+    T: RTreeObject,
+    Params: RTreeParams,
+    R: SelectionFunction<T>,
+{
+    remove_recursive::<_, Params, _>(node, removal_function, true)
+}
+
+fn remove_recursive<T, Params, R>(
+    node: &mut ParentNode<T>,
+    removal_function: &R,
+    all_matches: bool,
+) -> Vec<T>
```

Comment:
The naming doesn't really matter for this file anyway because it is private API only.

Regarding `drain_filter`, I do think that it would certainly be possible to implement a similar function for RTree. I am hesitant though since it would most certainly involve unsafe code and the `drain_filter` for `Vec` isn't even stable yet.

Reply:
I have tried to come up with an implementation of a drain like API but didn't really succeed with anything that is reasonably efficient.

My approach was like this:

```rust
struct RemovingIterator<'a, T, R>
where
    T: RTreeObject,
    R: SelectionFunction<T>,
{
    node_stack: Vec<(&'a mut ParentNode<T>, usize)>,
    selection_function: R,
}
```

where node_stack stores references to the various parent nodes currently in use along with a usize that points to the child of that node that is currently being investigated. However, I didn't manage to easily implement Iterator on that due to lifetime issues.

A simple alternative would be to use node_stack: Vec<usize> that only stores indices and needs to restore the current head node on each iteration. However, that appears to be overly inefficient, I'd rather suggest the Vec approach then.

Considering all of that, I would be happy with the Vec variant unless someone has an idea on how to implement the drain-like API!

Reply:
Yes, it would indeed be quite nice to have a `drain` like interface as it simplifies the API:  flexibility of removing a fixed number of items, and/or not having to collect it to a `Vec`.

Unfortunately, I believe the recursive stack can't be handled without the use of unsafe raw ptrs.  For instance, `BTreeMap` uses [a struct](https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#438-440) for iteration [that internally](https://doc.rust-lang.org/src/alloc/collections/btree/node.rs.html#182-193) uses `NonNull` or `*mut` to get rid of the lifetime issues.

As a compromise, should we make the `SelectionFunction` trait to take `&mut self` instead of `&self` ?  We don't really use it in parallel, so `&mut self` makes it a bit more flexible as it can store state within the function body (like how many have been selected).

Reply:
Investigating a bit more:  it seems that issue is that in `&mut T`, `T` is invariant, but in `&T` it is co-variant.  This seems to somehow prevent us from unwrapping the recursion in an iterator struct.  FWIW, here is a minimal iterator impl (uses `unsafe`):

```rust
struct DrainIterator<'a, T, R>
where
    T: RTreeObject,
    R: SelectionFunction<T>,
{
    node_stack: Vec<(*mut ParentNode<T>, usize)>,
    removal_function: R,
    _marker: PhantomData<&'a mut ()>,
}

impl<'a, T, R> Iterator for DrainIterator<'a, T, R>
where
    T: RTreeObject,
    R: SelectionFunction<T>,
{
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        let (node, idx) = match self.node_stack.last_mut() {
            Some((n, idx)) => (unsafe { &mut **n }, idx),
            None => return None,
        };

        if *idx > 0 || self.removal_function.should_unpack_parent(&node.envelope) {
            while *idx < node.children.len() {
                let child = &mut node.children[*idx];
                match child {
                    RTreeNode::Parent(ref mut data) => {
                        *idx += 1;
                        self.node_stack.push((data, 0));
                        return self.next();
                    }
                    RTreeNode::Leaf(ref b) => {
                        if self.removal_function.should_unpack_leaf(b) {
                            // Swap node with last, remove and return the value.
                            // No need to increment idx as something else has replaced it;
                            // or idx >= new len, and we'll handle it in the next iteration.
                            let data = match node.children.swap_remove(*idx) {
                                RTreeNode::Leaf(data) => data,
                                _ => unreachable!("RemovalIterator bug!"),
                            };

                            // As we've removed something, check if node is now empty
                            if node.children.is_empty() {
                                // Remove it from stack and remove the entry from parent node
                                self.node_stack.pop();
                                if let Some((node, parent_idx)) = self.node_stack.last() {
                                    let node = unsafe { &mut **node };
                                    assert!(*parent_idx > 1);
                                    node.children.swap_remove(*parent_idx - 1);
                                }
                            }

                            // As we've changed something compute envelopes again.
                            for (node, _) in self.node_stack.iter_mut().rev() {
                                let node = unsafe { &mut **node };
                                node.envelope = crate::node::envelope_for_children(&node.children);
                            }
                            return Some(data);
                        }
                    }
                }
            }
        }

        // Tail recursion for simplicity
        self.node_stack.pop();
        self.next()

    }
}
```

I quite like the flexibility it would provide.  However, given the complexity, and use of `unsafe`, I'm also inclined to just stick with the current impl. in this PR.

Reply:
Dug a bit deeper, trying to see how one would optimize the repeated compute of envelopes in the iterator version above.  I noticed that the iterator of `Vec::drain` (and `drain_filter`) actually reduce the length of the vector before iteration.  Then, in the `Drop` impl. of the iterator, they assemble it back (refer [here](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#1781)).   We could use a similar idea to optimize and remove all unsafe usage!

See a impl. sketch below; the key remarks are:
- Move the root `ParentNode` out of the `RTree` when the iterator is created
- Move child and put it back in during recursion.  This uses the fact that children of a `ParentNode` can be freely re-arranged.
- Compute the envelope as and when we pop things out of the stack
- No use of `unsafe` (but does involve [leak amplification](https://doc.rust-lang.org/nomicon/leaking.html#drain))

``` rust
struct DrainIterator<'a, T, R, Params>
where
    T: RTreeObject,
    Params: RTreeParams,
    R: SelectionFunction<T>,
{
    node_stack: Vec<(ParentNode<T>, usize)>,
    removal_function: R,
    rtree: &'a mut RTree<T, Params>,
}

impl<'a, T, R, Params> Iterator for DrainIterator<'a, T, R, Params>
where
    T: RTreeObject,
    Params: RTreeParams,
    R: SelectionFunction<T>,
{
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        let (node, idx) = match self.node_stack.last_mut() {
            Some(node) => (&mut node.0, &mut node.1),
            None => return None,
        };

        if *idx > 0 || self.removal_function.should_unpack_parent(&node.envelope) {
            while *idx < node.children.len() {
                match &mut node.children[*idx] {
                    RTreeNode::Parent(_) => {
                        // Swap node with last, remove and return the value.
                        // No need to increment idx as something else has replaced it;
                        // or idx == new len, and we'll handle it in the next iteration.
                        let child = match node.children.swap_remove(*idx) {
                            RTreeNode::Leaf(_) => unreachable!("DrainIterator bug!"),
                            RTreeNode::Parent(node) => node,
                        };
                        self.node_stack.push((child, 0));
                        return self.next();
                    }
                    RTreeNode::Leaf(ref leaf) => {
                        if self.removal_function.should_unpack_leaf(leaf) {
                            // Swap node with last, remove and return the value.
                            // No need to increment idx as something else has replaced it;
                            // or idx == new len, and we'll handle it in the next iteration.
                            return match node.children.swap_remove(*idx) {
                                RTreeNode::Leaf(data) => Some(data),
                                _ => unreachable!("RemovalIterator bug!"),
                            };
                        }
                        *idx += 1;
                    }
                }
            }
        }

        // At this point, the current node has been traversed.
        // Compute envelope for this node and push it back into the parent.
        // If the node is now empty, we can simply ignore it.
        let (mut node, _) = self.node_stack.pop().unwrap();
        if !node.children.is_empty() {

            // We only compute envelope for the current node as the parent
            // is taken care of when it is popped.

            // TODO: May be make this a method on `ParentNode`
            node.envelope = crate::node::envelope_for_children(&node.children);

            // Get the new top in stack
            let (parent_node, parent_idx) = match self.node_stack.last_mut() {
                Some(node) => (&mut node.0, &mut node.1),
                None => return None,
            };

            // Put the child back (but re-arranged)
            parent_node.children.push(RTreeNode::Parent(node));
            // Swap it with the current item and increment idx

            // Note that during iteration, parent_idx may be equal to
            // children.len(), but this is okay as the swap will be a no-op.
            let parent_len = parent_node.children.len();
            parent_node.children.swap(*parent_idx, parent_len - 1);
            *parent_idx += 1;
        }
        self.next()
    }
}

impl<'a, T, R, Params> Drop for DrainIterator<'a, T, R, Params>
where
    T: RTreeObject,
    Params: RTreeParams,
    R: SelectionFunction<T>,
{
    fn drop(&mut self) {
        // Re-assemble back the original rtree
        // Compute envelope as we re-assemble.

        // Basically the last section of the Iterator::next impl above, but
        // until the stack is empty.
        todo!()
    }
}
```

This has all the advantages of a draining iterator, but doesn't need many allocations.  The only allocation is the `node_stack`, which could also be optimized if we know the height of the RTree to start with.  It does need ~O(n) more copies of a `ParentNode` which roughly amounts to a pointer + envelope, but doesn't affect any asymptotics.

One other advantage is that in the future, we could even consider re-balancing the RTree in the destructor.  Currently, all the remove variants could theoretically skew the tree so that the O(log size-of-tree) bound on search is no longer true assuming we insert a lot of entries and remove most of them.

PS: Apologies for the message spam; I just wanted to brainstorm a bit on an intuitive and flexible API for the removals.

Reply:
This is some non-trivial code, but _wow_, what a cool use of mutable lifetimes + Drop! Thanks for prototyping this @rmanoka, I learned some things.